Promise support.

kyptov opened the pull request:
If callback missing than method will return promise. By default using native Promise. 
I tried to make minimal changes to preserve backwards compatibility. #37 

Some methods throw error. I didn't change anything in it. Better to use methods from promise chain to catch those errors..

```
Promise.resolve().then(function () {
    return jira.myself.getMyself();
}).then(function (data) {
    console.log(data);
}).catch(function (err) {
    console.log(err);
});
```

JonahTurnquist commented:
+1

Also I wouldn't care if callback support was dropped.. but it would have to be a major version update

floralvikings commented:
To be clear: callback support will **not** be dropped from this package.  There is just too much existing code to break backwards compatibility in such a major fashion, and I have no interest in maintaining two major version paths.  If you'd prefer an API that uses promises exclusively, you're more than welcome to fork this repository and release (and maintain) one.  I'd be happy to link to it from this repository; more options make life better for everyone (except the maintainer 😝 ).  Just leave a comment on this PR or #37.

*That being said*, at a glance this looks like a very good compromise and doesn't seem to add much burden to further contributions or support.  I'll try to do a more thorough review and test this sometime this week, but it's a pretty large change and I find myself very short on free time lately (which is why I never got around to reviewing #49), so no promises (pun intended 😄 ).

kyptov commented:
I will briefly explain changes:

### 1. 
in 'api' folder there is a lot of scripts making request with 
`this.jiraClient.makeRequest()`, 
I just added **return** to it.
`return this.jiraClient.makeRequest()`

That's it, no more no less. Also updated jsdoc, but it can not break code. 
If you using callback, method will return undefined in fork version as in master version.

### 2.
index.js in constructor added extra parameter `config.promise` for those who wanted any specific promise library to use, by default it use native Promise (can be problem in node < 0.12). If this is a problem I can add additional checking like: 
```
if (config.promise) {
     this.promise = config.promise || Promise;
}
```

### 3. 
in index.js in makeRequest() method I added additional condition
```
if (callback) {
// your code of request without changes
} else if (this.promise) {
// new code of request with promise
}
```

As you see to maintain version with promise support just need to add **return** in api methods.
